# Coursera Courses

In [ ]:
#Import all the libraries 
import pandas as pd
import numpy as np
import chart_studio.plotly as py
import cufflinks as cf
import seaborn as sns
import plotly.express as px
import matplotlib.pyplot as plt
from scipy import stats
%matplotlib inline

In [1]:
file_path = r'C:\Users\joaol\Desktop\Coursera_data\coursea_data.csv' #Use the correct path in your pc 
df = pd.read_csv(file_path)


NameError: name 'os' is not defined

In [ ]:
#To check the data set 
df.head()

Unnamed: 0                                       course_title  \
0         134  (ISC)² Systems Security Certified Practitioner...   
1         743  A Crash Course in Causality:  Inferring Causal...   
2         874                     A Crash Course in Data Science   
3         413                            A Law Student's Toolkit   
4         635                A Life of Happiness and Fulfillment   

          course_organization course_Certificate_type  course_rating  \
0                      (ISC)²          SPECIALIZATION            4.7   
1  University of Pennsylvania                  COURSE            4.7   
2    Johns Hopkins University                  COURSE            4.5   
3             Yale University                  COURSE            4.7   
4   Indian School of Business                  COURSE            4.8   

  course_difficulty course_students_enrolled  
0          Beginner                     5.3k  
1      Intermediate                      17k  
2             Mixed                     130k  
3             Mixed                      91k  
4             Mixed                     320k

# Hypothesis: 
### Courses that are Non-Beginner have greater impact in student's enrollment and course ratings

In [ ]:
#To check for the data types of the data set
print(df.dtypes)


Unnamed: 0                    int64
course_title                 object
course_organization          object
course_Certificate_type      object
course_rating               float64
course_difficulty            object
course_students_enrolled     object
dtype: object


In [ ]:
#To check for missing values 
print(df.isnull().sum())


Unnamed: 0                  0
course_title                0
course_organization         0
course_Certificate_type     0
course_rating               0
course_difficulty           0
course_students_enrolled    0
dtype: int64


In [ ]:
#To check for duplicated values
print(df.duplicated().sum())

0


In [ ]:
#To check for duplicated values in specific columns
duplicates = df['course_title'].value_counts()
print(duplicates[duplicates > 1])


course_title
Developing Your Musicianship    2
Marketing Digital               2
Machine Learning                2
Name: count, dtype: int64


In [ ]:
#Chnage the name of the column to an ID column
df.rename(columns={'Unnamed: 0': 'course_id'}, 
          inplace=True)
df.head()


course_id                                       course_title  \
0        134  (ISC)² Systems Security Certified Practitioner...   
1        743  A Crash Course in Causality:  Inferring Causal...   
2        874                     A Crash Course in Data Science   
3        413                            A Law Student's Toolkit   
4        635                A Life of Happiness and Fulfillment   

          course_organization course_Certificate_type  course_rating  \
0                      (ISC)²          SPECIALIZATION            4.7   
1  University of Pennsylvania                  COURSE            4.7   
2    Johns Hopkins University                  COURSE            4.5   
3             Yale University                  COURSE            4.7   
4   Indian School of Business                  COURSE            4.8   

  course_difficulty course_students_enrolled  
0          Beginner                     5.3k  
1      Intermediate                      17k  
2             Mixed                     130k  
3             Mixed                      91k  
4             Mixed                     320k

In [ ]:
#To check for how the nº of students enrolled has been put 
df['course_students_enrolled'].unique()

array(['5.3k', '17k', '130k', '91k', '320k', '39k', '350k', '2.4k', '61k',
       '12k', '4k', '13k', '11k', '27k', '110k', '6.6k', '540k', '22k',
       '2.9k', '80k', '9.9k', '23k', '9.2k', '78k', '190k', '35k', '29k',
       '14k', '21k', '94k', '69k', '40k', '220k', '150k', '18k', '270k',
       '7.9k', '30k', '36k', '20k', '8.1k', '120k', '71k', '63k', '42k',
       '480k', '97k', '200k', '180k', '4.2k', '310k', '3.9k', '79k',
       '31k', '15k', '10k', '66k', '33k', '56k', '7.3k', '9.7k', '210k',
       '28k', '6.5k', '55k', '2.3k', '8.8k', '88k', '1.9k', '68k', '1.6k',
       '9.5k', '57k', '26k', '84k', '95k', '5.8k', '24k', '67k', '280k',
       '38k', '77k', '510k', '89k', '48k', '160k', '32k', '340k', '82k',
       '790k', '19k', '51k', '4.8k', '37k', '43k', '6.4k', '49k', '240k',
       '46k', '7.8k', '75k', '81k', '140k', '5.5k', '99k', '100k', '3k',
       '830k', '740k', '60k', '96k', '690k', '44k', '4.5k', '8.2k', '16k',
       '300k', '8k', '41k', '54k', '9k', '380k',

In [ ]:
#To change the values of K to thousands and M to millions
def n_students(x):
    if isinstance(x, str):  
        if x[-1] == 'k':
            return int(float(x.replace('k', '')) * 1000)  
        elif x[-1] == 'm':
            return int(float(x.replace('m', '')) * 1000000)
        else:
            return int(float(x))  
    else:  
        return int(x)

df['course_students_enrolled'] = df['course_students_enrolled'].apply(n_students)
df.head()  

course_id                                       course_title  \
0        134  (ISC)² Systems Security Certified Practitioner...   
1        743  A Crash Course in Causality:  Inferring Causal...   
2        874                     A Crash Course in Data Science   
3        413                            A Law Student's Toolkit   
4        635                A Life of Happiness and Fulfillment   

          course_organization course_Certificate_type  course_rating  \
0                      (ISC)²          SPECIALIZATION            4.7   
1  University of Pennsylvania                  COURSE            4.7   
2    Johns Hopkins University                  COURSE            4.5   
3             Yale University                  COURSE            4.7   
4   Indian School of Business                  COURSE            4.8   

  course_difficulty  course_students_enrolled  
0          Beginner                      5300  
1      Intermediate                     17000  
2             Mixed                    130000  
3             Mixed                     91000  
4             Mixed                    320000

In [ ]:
#To check for the nº of rows in the data set
df['course_id'].count()

891

In [ ]:
#To check for the different types of certificates
df['course_Certificate_type'].value_counts()

course_Certificate_type
COURSE                      582
SPECIALIZATION              297
PROFESSIONAL CERTIFICATE     12
Name: count, dtype: int64

In [ ]:
#To check for different types of difficulties
df['course_difficulty'].value_counts()

course_difficulty
Beginner        487
Intermediate    198
Mixed           187
Advanced         19
Name: count, dtype: int64

In [124]:
#To join non-beginners levels of difficulties to compare it with beginner and non-beginner courses
df['difficulty_groups'] = df['course_difficulty'].apply(
    lambda x: 'Beginner' if x == 'Beginner' else 'Non-beginner')

difficulty_groups = df.groupby('difficulty_groups').agg(
    {'course_students_enrolled': 'mean',
     'course_rating': 'mean'}
     ).reset_index()

difficulty_groups = difficulty_groups.round({'course_students_enrolled': 2, 'course_rating': 2})
print(difficulty_groups)
df.groupby('difficulty_groups')['course_students_enrolled'].describe()

  difficulty_groups  course_students_enrolled  course_rating
0          Beginner                  81974.95           4.68
1      Non-beginner                 100891.34           4.67


count           mean            std     min      25%  \
difficulty_groups                                                         
Beginner           487.0   81974.948665  127270.770733  1500.0  15000.0   
Non-beginner       404.0  100891.336634  231022.152159  1600.0  21000.0   

                       50%       75%        max  
difficulty_groups                                
Beginner           37000.0   98500.0  1500000.0  
Non-beginner       47000.0  100000.0  3200000.0

In [125]:
df.groupby('difficulty_groups')['course_rating'].describe()

count      mean       std  min  25%  50%  75%  max
difficulty_groups                                                    
Beginner           487.0  4.680903  0.141947  4.1  4.6  4.7  4.8  5.0
Non-beginner       404.0  4.673020  0.183811  3.3  4.6  4.7  4.8  5.0

# Hypothesis considerations:
- In this dataset, there are 891 different courses
- There are 4 different difficulty levels:
    - Beginner
    - Intermediate 
    - Mixed 
    - Advanced 
- The Non-beginner course are the ones in the intermediate, mixed and advanced levels. 
- There are 404 non-beginner courses and 487 beginner courses.


In [127]:
#With this code, I will look for the nº of outliers in beginners and non-beginners courses
Q1 = df['course_students_enrolled'].quantile(0.25)
Q3 = df['course_students_enrolled'].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

beginner_outliers = df[(df['difficulty_groups'] == 'Beginner') &
                       ((df['course_students_enrolled'] < lower_bound) |
                        (df['course_students_enrolled'] > upper_bound))]

non_beginner_outliers = df[(df['difficulty_groups'] == 'Non-beginner') &
                           ((df['course_students_enrolled'] < lower_bound) |
                            (df['course_students_enrolled'] > upper_bound))]

n_beginner_outliers = beginner_outliers.shape[0]
n_non_beginner_outliers = non_beginner_outliers.shape[0]


print(f"Number of outliers in Beginner group: {n_beginner_outliers}")
print(f"Number of outliers in Non-beginner group: {n_non_beginner_outliers}")

Number of outliers in Beginner group: 42
Number of outliers in Non-beginner group: 36


In [128]:
#To check for distributions in beginner and  non-beginner courses
fig = px.box(df, 
             x='difficulty_groups', 
             y='course_students_enrolled', 
             points='suspectedoutliers', 
             title='Distribution of Students Enrolled by Difficulty Group',
             labels={'difficulty_groups': 'Difficulty', 'course_students_enrolled': 'Nº of students enrolled'},
             color='difficulty_groups',
             color_discrete_map={'Beginner': '#008080', 'Non-beginner': '#808000'})  

fig.update_yaxes(range=[0,230000])

fig.update_layout(
    paper_bgcolor='rgb(250, 250, 238)',  
    plot_bgcolor='rgb(250, 250, 238)')  

fig.show()



In [129]:
#To check for distributions in beginner and non-beginner courses but with outliers 
fig = px.violin(df, 
             x='difficulty_groups', 
             y='course_students_enrolled',  
             title='Distribution of Students Enrolled by Difficulty Group',
             labels={'difficulty_groups': 'Difficulty', 'course_students_enrolled': 'Nº of students enrolled'},
             color='difficulty_groups',
             color_discrete_map={'Beginner': '#008080', 'Non-beginner': '#808000'},
             box=True,
             points= 'all')  

fig.update_layout(
    paper_bgcolor='rgb(250, 250, 238)',  
    plot_bgcolor='rgb(250, 250, 238)')  

fig.show()

In [130]:
#To check for the mean of students enrolled in beginners and non-beginners courses
#The descending function was to make sure that the columns are always on the same side
difficulty_groups_des = difficulty_groups.sort_values(by='course_rating', ascending=False)

fig=px.bar(difficulty_groups_des, 
           x='difficulty_groups', 
           y='course_students_enrolled',
           title='Mean of students enrolled in Beginer vs Non-beginer',
           labels={'difficulty_groups': 'Difficulty', 'course_students_enrolled': 'Nº of students enrolled'})

fig.update_yaxes(range=[0, 120000])  
fig.update_traces(text=difficulty_groups_des['course_students_enrolled'],  
                  textposition='outside',  
                  marker=dict(color=['#008080', '#808000']))

fig.update_layout(paper_bgcolor='rgb(250, 250, 238)', plot_bgcolor='rgb(250, 250, 238)') 


fig.show()

In [155]:
#The pie chart is for us to see if there's a difference in the mean of ratings for beginners and non-begginers

fig = px.pie(difficulty_groups,
    values='course_rating',
    names='difficulty_groups',
    title='Course ratings by difficulties')

fig.update_traces(textinfo='label+percent+value',  
    textposition='inside',
    marker=dict(colors=['#008080','#808000'])  
)
fig.update_layout(paper_bgcolor='rgb(250, 250, 238)', )

fig.show()

In [157]:
# Check for the rating being impacted by students enrolled, 3 features: beginner/non-beginner, rating, students enrolled
fig = px.scatter(
    df,
    x='course_students_enrolled',
    y='course_rating',
    color='difficulty_groups',
    title='Impact of ratings on rtudents enrolled by difficulty',
    labels={'difficulty_groups': 'Difficulty', 
            'course_students_enrolled': 'Nº of students enrolled', 
            'course_rating': 'Rating'},
    size='course_students_enrolled', 
    hover_data=['course_title',
                'course_organization'],
    color_discrete_map={
        'Beginner': '#008080', 
        'Non-beginner': '#808000'})     



fig.update_layout(paper_bgcolor='rgb(250, 250, 238)',  
                  plot_bgcolor='rgb(250, 250, 238)')

fig.show()


In [156]:
fig = px.scatter(
    df,
    x='course_students_enrolled',
    y='course_rating',
    color='difficulty_groups',
    facet_col='difficulty_groups',  
    title='Impact of ratings on students enrolled by difficulty',
    labels={'difficulty_groups': 'Difficulty',
            'course_students_enrolled': 'Nº of students enrolled', 
            'course_rating': 'Rating'},
    hover_data=['course_title', 
                'course_organization'],
    color_discrete_map={'Beginner': '#008080', 
                        'Non-beginner': '#808000'})

fig.update_layout(paper_bgcolor='rgb(250, 250, 238)',  
    plot_bgcolor='rgb(250, 250, 238)')

fig.show()

In [137]:
#To check the top 5 courses in the data set, according to the students enrolled
best_ranked_courses = df.sort_values(by='course_students_enrolled', 
                                     ascending=False).head(5)

print("The 5 best-ranked courses are:")
for index, row in best_ranked_courses.iterrows():
    print(f"Course Name: {row['course_title']} | Students Enrolled: {row['course_students_enrolled']} | Rating: {row['course_rating']} | Difficulty: {row['difficulty_groups']}")

                     

The 5 best-ranked courses are:
Course Name: Machine Learning | Students Enrolled: 3200000 | Rating: 4.9 | Difficulty: Non-beginner
Course Name: The Science of Well-Being | Students Enrolled: 2500000 | Rating: 4.9 | Difficulty: Non-beginner
Course Name: Python for Everybody | Students Enrolled: 1500000 | Rating: 4.8 | Difficulty: Beginner
Course Name: Programming for Everybody (Getting Started with Python) | Students Enrolled: 1300000 | Rating: 4.8 | Difficulty: Non-beginner
Course Name: Data Science | Students Enrolled: 830000 | Rating: 4.5 | Difficulty: Beginner


In [ ]:
#T-test for course ratings between beginner and non beginner courses
beginner_ratings = df[df['difficulty_groups'] == 'beginner']['course_rating']
non_beginner_ratings = df[df['difficulty_groups'] == 'Non-beginner']['course_rating']

t_stat, p_value = stats.ttest_ind(beginner_ratings,
                                  non_beginner_ratings)


t_stat_round = round(t_stat, 4)
p_value_round = round(p_value, 4)


print(f"T-statistic: {t_stat_round}")
print(f"P-value: {p_value_round}")



T-statistic: 0.722
P-value: 0.4705


In [ ]:
#T-test for students enrolled between beginner and non beginner courses
beginner_students = df[df['difficulty_groups'] == 'beginner']['course_students_enrolled']
non_beginner_students = df[df['difficulty_groups'] == 'Non-beginner']['course_students_enrolled']

t_stat, p_value = stats.ttest_ind(beginner_students,
                                  non_beginner_students)

t_stat_round = round(t_stat, 4)
p_value_round = round(p_value, 4)

print(f"T-statistic: {t_stat_round}")
print(f"P-value: {p_value_round}")

T-statistic: -1.5462
P-value: 0.1224


# Conclusion:
##### With all the data and t-test completed, we cannot reject the null hypothesis, and come to the conclusion that there’s no statistical difference between beginner and non-beginner courses regarding to their popularity or their rating.


## Ways to expand more this analysis
- Can check if type of certification has an impact on the students enrolled and course rating
- Are there any particular organizations that have better ratings compared to others
- Look deeper into course themes that can be more popular in this data set